zabbix_auto_config/models.py

```diff
@@ -0,0 +1,105 @@
+import logging
+
+from typing import (
+    Dict,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+from pydantic import (
+    BaseModel,
+    conint,
+    validator,
+)
+
+from . import utils
+
+# TODO: Models aren't validated when making changes to a set/list. Why? How to handle?
```

Comment:
Do you have an example that fails?  Inspired by [this article](https://medium.com/swlh/cool-things-you-can-do-with-pydantic-fc1c948fbde0), I played around a bit to try and reproduce:

```
from pydantic import BaseModel, ValidationError
from typing import (Optional, Set)
import pytest

class Host(BaseModel):
    enabled: bool
    properties: Optional[Set[str]] = set()

    class Config:
        validate_assignment = True

def test_pydantic_host():
    host = Host(properties=["foo=bar"], enabled=True)
    print("{}".format(host))
    assert host.properties == {"foo=bar"}
    host.properties = [1, 2]
    print("{}".format(host))
    host.properties = ["foo", [1, 2]]
    print("{}".format(host))

test_pydantic_host()
```

I found that pydantic will implicitly cast values to the declared type, i.e. `host.properties = [1, 2]` becomes `{'1', '2'}`; whereas `host.properties = ["foo", [1, 2]]` raises the expected `ValidationError`.  So AFAICT this comment can be removed.

Reply:
Assignments are validated, but mutations are not. A simple example:
```python
from typing import List, Set

import pydantic

class A(pydantic.BaseModel):
    b: List[pydantic.conint(ge=0)] = []
    c: Set[pydantic.conint(ge=0)] = set()

    class Config:
        validate_assignment = True

a = A()
a.b.append(-1)  # no error
a.b.append("a")  # no error
try:
  a.b = [-1, "a"]  # error
except pydantic.error_wrappers.ValidationError as e:
    print("Error:", e)

a.c.add(-1)  # no error
a.c.add("a")  # no error
try:
  a.c = {-1, "a"}  # error
except pydantic.error_wrappers.ValidationError as e:
    print("Error:", e)

print(a)
```

I'm leaning towards that this is OK for now. Host modifiers and collectors are/should be somewhat trusted code. The validator is more a convenience than a strict "firewall". But I'm happy to hear more input?

If the validator should be more strict I guess we'll have to create a wrapping class around the lists and sets(?)

PS. I haven't read that article, but I'll do it when I have some spare time 👍 I don't have much experience with pydantic, but I'm happy to learn more.

Reply:
Ooh, I see, thanks for the example.  I did not consider object mutations (still a Python noob!).  Don't get too excited about the article, it was just the first hit on a random web search...

Thinking out loud (and getting frustrated by even `allow_mutation` not preventing this problem), one thing that has been on the back of my mind is enabling host modifiers and source collectors written in other programming languages.  That's because zac could be useful in another project I'm involved with, but they are not too fond of Python...

(please excuse my daydreaming)

To achieve that, I imagine a JSON-based protocol where the host modifier gets the current state on standard input, and outputs a modified state, which is then validated and sent to Zabbix (or the next modifier).

Devising such a protocol would make validation only occur on the various JSON responses, instead of letting "random" plugins muck about with shared state.

Do you think designing a protocol for source collectors and host modifiers is feasible?

Reply:
Yeah, I think it's a bit weird that you're allowed to do stuff that makes the object not conform with the class/model.

It's a pro/con situation. Should zac be more strict (and probably complicated?) or should it trust the user not to do weird stuff. I'm currently leaning towards the latter.

That other part: I'm not sure I fully follow. At some point the python code will need to be able to communicate with some other part. I guess you could make source collectors/host modifiers "wrap" some other binary/script by using something like [subprocess.run](https://docs.python.org/3/library/subprocess.html#subprocess.run)? My first thought is that the host modifiers will produce some (a lot?) of overhead 🤔 I'd be happy to discuss the matter, but maybe outside of this PR?